sdks/java/io/mqtt/src/main/java/org/apache/beam/sdk/io/mqtt/MqttIO.java

```diff
@@ -0,0 +1,518 @@
+ *
+ * pipeline
+ *   .apply(...) // provide PCollection<byte[]>
+ *   .MqttIO.write()
+ *     .withMqttConnectionConfiguration(MqttIO.MqttConnectionConfiguration.create(
+ *       "tcp://host:11883",
+ *       "my_client_id",
+ *       "my_topic"))
+ *     .withRetained(true)
+ *     .withQoS(2)
+ *
+ * }</pre>
+ */
+public class MqttIO {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(MqttIO.class);
+
+  public static Read read() {
+    return new AutoValue_MqttIO_Read.Builder()
+        .setMaxReadTime(null).setMaxNumRecords(Long.MAX_VALUE).build();
+  }
+
+  public static Write write() {
+    return new AutoValue_MqttIO_Write.Builder()
+        .setRetained(false)
+        .setQos(0)
+        .build();
+  }
+
+  private MqttIO() {
+  }
+
+  /**
+   * A POJO describing a MQTT connection.
+   */
+  @AutoValue
+  abstract static class MqttConnectionConfiguration implements Serializable {
+
+    @Nullable abstract String serverUri();
+    @Nullable abstract String clientId();
+    @Nullable abstract String topic();
+
+    public static MqttConnectionConfiguration create(String serverUri, String clientId,
+                                                     String topic) {
+      checkNotNull(serverUri, "serverUri");
+      checkNotNull(clientId, "clientId");
+      checkNotNull(topic, "topic");
+      return new AutoValue_MqttIO_MqttConnectionConfiguration(serverUri, clientId, topic);
+    }
+
+    private void validate() {
+      checkNotNull(serverUri());
+      checkNotNull(clientId());
+    }
+
+    private void populateDisplayData(DisplayData.Builder builder) {
+      builder.add(DisplayData.item("serverUri", serverUri()));
+      builder.add(DisplayData.item("clientId", clientId()));
+      builder.add(DisplayData.item("topic", topic()));
+    }
+
+    MqttClient getClient() throws Exception {
+      MqttClient client = new MqttClient(serverUri(), clientId());
+      client.connect();
+      return client;
+    }
+
+  }
+
+  /**
+   * A {@link PTransform} to read from a MQTT broker.
+   */
+  @AutoValue
+  public abstract static class Read extends PTransform<PBegin, PCollection<byte[]>> {
+
+    @Nullable abstract MqttConnectionConfiguration mqttConnectionConfiguration();
+    @Nullable abstract long maxNumRecords();
+    @Nullable abstract Duration maxReadTime();
+
+    abstract Builder toBuilder();
+
+    @AutoValue.Builder
+    abstract static class Builder {
+      abstract Builder setMqttConnectionConfiguration(MqttConnectionConfiguration config);
+      abstract Builder setMaxNumRecords(long maxNumRecords);
+      abstract Builder setMaxReadTime(Duration maxReadTime);
+      abstract Read build();
+    }
+
+    public Read withMqttConnectionConfiguration(MqttConnectionConfiguration configuration) {
+      checkNotNull(configuration, "MqttConnectionConfiguration");
+      return toBuilder().setMqttConnectionConfiguration(configuration).build();
+    }
+
+    public Read withMaxNumRecords(long maxNumRecords) {
```

Comment:
Either disallow setting both, or specify which takes priority between maxNumRecords and maxReadTime.